src/internal/m365/onedrive/permission.go

```diff
@@ -344,36 +418,52 @@ func RestorePermissions(
 
 	ctx = clues.Add(ctx, "permission_item_id", itemID)
 
-	previous, err := computePreviousMetadata(ctx, itemPath, caches.ParentDirToMeta)
+	previousLinkShares, err := computePreviousLinkShares(ctx, itemPath, caches.ParentDirToMeta)
 	if err != nil {
-		return clues.Wrap(err, "previous metadata")
+		return clues.Wrap(err, "previous link shares")
 	}
 
-	permAdded, permRemoved := metadata.DiffPermissions(previous.Permissions, current.Permissions)
-	lsAdded, lsRemoved := metadata.DiffPermissions(previous.LinkShares, current.LinkShares)
+	lsAdded, lsRemoved := metadata.DiffLinkShares(previousLinkShares, current.LinkShares)
 
-	err = UpdatePermissions(
+	didReset, err := UpdateLinkShares(
 		ctx,
 		rh,
 		driveID,
 		itemID,
-		permAdded,
-		permRemoved,
-		caches.OldPermIDToNewID)
+		lsAdded,
+		lsRemoved,
+		caches.OldLinkShareIDToNewID)
 	if err != nil {
-		return clues.Wrap(err, "updating permissions")
+		return clues.Wrap(err, "updating link shares")
 	}
 
-	err = UpdateLinkShares(
+	permRemoved := []metadata.Permission{}
+	permAdded := current.Permissions
+
+	if !didReset {
+		// In case we did a reset of permissions when restoring link
+		// shares, we have to make sure to restore all the permissions
+		// that an item has as they too will be removed.
```

Comment:
this comment seems at odds with the if-condition. Should the if-condition be flipped?

Reply:
Ahh, missed updating the code after flipping the condition. I did this to avoid calling `computePreviousMetadata` and `DiffPermissions` if not needed, but I think they are not that expensive so flipped it back. I think the code reads better this way.